emma-language/src/test/scala/org/emmalanguage/compiler/lang/backend/CachingSpec.scala

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright © 2014 TU Berlin (emma@dima.tu-berlin.de)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.emmalanguage
+package compiler.lang.backend
+
+import compiler.ir
+
+import api.DataBag
+import compiler.BaseCompilerSpec
+
+/** A spec for order disambiguation. */
+class CachingSpec extends BaseCompilerSpec {
+
+  import compiler._
+
+  val liftPipeline: u.Expr[Any] => u.Tree =
+    pipeline(typeCheck = true)(
+      Core.lift
+    ).compose(_.tree)
+
+  val addCacheCalls: u.Expr[Any] => u.Tree =
+    pipeline(typeCheck = true)(
+      Core.lift,
+      tree => time(Caching.addCacheCalls(tree), "addCacheCalls")
+    ).compose(_.tree)
+
+  "cache " - {
+
+    "DataBag terms used multiple times" in {
+
+      val inp = u.reify {
+        val i = 2
+        val xs = DataBag(Seq(1, 2, 3, 4, 5))
+
+        val ys = xs.map(_ + i)
+        val zs = xs.map(_ * i)
+
+        ys union zs
+      }
+
+      val exp = u.reify {
+        val i = 2
+        val xs = ir.Runtime.cache(DataBag(Seq(1, 2, 3, 4, 5)))
+
+        val ys = xs.map(_ + i)
+        val zs = xs.map(_ * i)
+
+        ys union zs
+      }
+
+      addCacheCalls(inp) shouldBe alphaEqTo(liftPipeline(exp))
+    }
+
+    "DataBag terms on a dynamic path (loop)" in {
+
+      val inp = u.reify {
+        val i = 2
+        val N = 5
+        var xs = DataBag(Seq(1, 2, 3, 4, 5))
+
+        for (_ <- 0 to N)
+          xs = xs.map(_ + i)
+
+        xs
+      }
+
+      val exp = u.reify {
+        val i = 2
+        val N = 5
+        var xs = ir.Runtime.cache(DataBag(Seq(1, 2, 3, 4, 5)))
```

Comment:
@joroKr21 The point is to avoid constructing large dataflows:

in the spec, example if we have `val N = 100`, this will yield a single (non-materialized) dataflow with a chain of 100 `map` functions. In cases where the unrolled dataflow structure is non-linear the underlying engine might break.